envoy/network/dns_resolver.h

```diff
@@ -26,7 +26,30 @@ class DnsResolverFactory : public Config::TypedFactory {
       const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config) const PURE;
 
   std::string category() const override { return std::string(DnsResolverCategory); }
+
+  /**
+   * Initialize the related data for this type of DNS resolver.
+   * For some DNS resolvers, like c-ares, there are some specific data structure
+   * needs to be initialized before using it to resolve target.
+   */
+  virtual void initialize() {}
+
+  /**
+   * Cleanup the related data for this type of DNS resolver.
+   * For some DNS resolvers, like c-ares, there are some specific data structure
+   * needs to be cleaned up before terminates Envoy.
+   */
+  virtual void terminate() {}
 };
 
+/**
+ * Terminate the DNS resolver factories.
+ */
+static inline void terminateDnsResolverFactories() {
+  auto& factories = Registry::FactoryRegistry<Network::DnsResolverFactory>::factories();
+  std::for_each(factories.begin(), factories.end(),
+                [](auto& factory_it) { factory_it.second->terminate(); });
+}
+
```

Comment:
this looks reasonable. So that the abstraction is somewhat more symmetric can you declare a function `intializeDnsResolverFactory(const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config)` and put it here, to abstract the call to `Config::Utility::getAndCheckFactory<Network::DnsResolverFactory>(typed_dns_resolver_config);` and `factory.initialize()` ? The impl would also go in dns_resolver_util.cc.